rs/sns/governance/src/governance.rs

```diff
@@ -5030,13 +5013,41 @@ impl Governance {
         }
     }
 
-    /// Invalidates the cached upgrade steps.
-    fn invalidate_cached_upgrade_steps(&mut self, human_readable: String) {
+    /// Returns the upgrade steps that are guaranteed to start from `current_version`, resetting
+    /// `cached_upgrade_steps` and `target_version` if an inconsistency is detected.
+    pub(crate) fn get_or_reset_upgrade_steps(
+        &mut self,
+        current_version: &Version,
+    ) -> CachedUpgradeSteps {
+        let human_readable = match self
+            .proto
+            .cached_upgrade_steps_or_err()
+            .and_then(|cached_upgrade_steps| cached_upgrade_steps.take_from(current_version))
+        {
+            Ok(upgrade_steps) => {
+                // Happy case.
+                return upgrade_steps;
+            }
+            Err(err) => err,
+        };
+
         self.push_to_upgrade_journal(upgrade_journal_entry::UpgradeStepsReset::new(
-            human_readable,
-            vec![],
+            human_readable.clone(),
+            vec![current_version.clone()],
         ));
-        self.proto.cached_upgrade_steps = None;
+
+        let cached_upgrade_steps =
+            CachedUpgradeSteps::without_pending_upgrades(current_version.clone(), self.env.now());
+
+        self.proto
+            .cached_upgrade_steps
+            .replace(CachedUpgradeStepsPb::from(cached_upgrade_steps.clone()));
+
+        if self.proto.target_version.is_some() {
+            self.invalidate_target_version(human_readable)
+        }
+
+        cached_upgrade_steps
     }
```

Comment:
Would be nice to have a test that after get_or_reset_upgrade_steps resets the upgrade steps, should_refresh_cached_upgrade_steps() starts returning true

Reply:
Added get_or_reset_upgrade_steps_leads_to_should_refresh_cached_upgrade_steps